libr/util/event.c

```diff
@@ -34,50 +37,60 @@ R_API void r_event_free(REvent *ev) {
 	free (ev);
 }
 
-static bool add_hook(void *cb, const ut64 k, const void *v) {
-	RPVector *cbs = (RPVector *)v;
+static bool add_hook(void *hook, const ut64 k, const void *v) {
+	RVector *cbs = (RVector *)v;
 	r_return_val_if_fail (cbs, false);
-	r_pvector_push (cbs, cb);
+	r_vector_push (cbs, hook);
 	return true;
 }
 
-R_API void r_event_hook(REvent *ev, int type, REventCallback cb) {
-	r_return_if_fail (ev);
+R_API int r_event_hook(REvent *ev, int type, REventCallback cb, void *user) {
+	r_return_val_if_fail (ev, -1);
+	REventCallbackHook hook;
+	hook.cb = cb;
+	hook.user = user;
+	hook.handle = ev->hook_handle_next++;
 	if (type == R_EVENT_ALL) {
-		ht_up_foreach (ev->callbacks, add_hook, cb);
+		ht_up_foreach (ev->callbacks, add_hook, &hook);
 	} else {
-		RPVector *cbs = ht_up_find (ev->callbacks, type, NULL);
-		add_hook (cb, 0, cbs);
+		RPVector *cbs = ht_up_find (ev->callbacks, (ut64)type, NULL);
+		add_hook (&hook, 0, cbs);
 	}
+	return hook.handle;
 }
 
-static bool del_hook(void *cb, const ut64 k, const void *v) {
-	RPVector *cbs = (RPVector *)v;
+static bool del_hook(void *user, const ut64 k, const void *v) {
+	int handle = (int)(intptr_t)user;
+	RVector *cbs = (RVector *)v;
 	r_return_val_if_fail (cbs, false);
-	r_pvector_remove_data (cbs, cb);
+	for (size_t i=0; i<cbs->len; i++) {
+		REventCallbackHook *hook = r_vector_index_ptr (cbs, i);
+		if (hook->handle == handle) {
+			r_vector_remove_at (cbs, i, NULL);
+			break;
+		}
+	}
 	return true;
 }
 
-R_API void r_event_unhook(REvent *ev, int type, REventCallback cb) {
+R_API void r_event_unhook(REvent *ev, int type, int handle) {
 	r_return_if_fail (ev);
 	if (type == R_EVENT_ALL) {
-		ht_up_foreach (ev->callbacks, del_hook, cb);
+		ht_up_foreach (ev->callbacks, del_hook, (void *)(intptr_t )handle);
```

Comment:
uhmf casting int to void* is ... fine on 32bits, but its a bad practice

Reply:
please just use `&handle` and get a `int *handle = (int *)user` in the `del_hook` function.